### PR TITLE
Conditionally use std::invoke_result_t

### PR DESCRIPTION
**Pull request by tsnorri**
- std::result_of has been deprecated in C++17 and removed in C++20.

### CODE REVIEW

**Comment by tsnorri** [comment]
Never mind the comment about non-optimality; [P0604R0](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0604r0.html) says `__cpp_lib_is_invocable` is the correct macro.

**Comment by eseiler** [comment]
Thanks for your contribution!

Since we aim at cpp17, I'm gonna go ahead and remove the check and just use the cpp17 version (and see if the CI accepts this).

File: include/sdsl/iterators.hpp [hunk 1]
```diff
@@ -139,7 +140,11 @@ struct random_access_container
 {
     typedef int_vector<>::size_type size_type;
     typedef int_vector<>::difference_type difference_type;
+#ifdef __cpp_lib_is_invocable
+    typedef typename std::invoke_result_t<t_F, size_type> value_type;
+#else
     typedef typename std::result_of<t_F(size_type)>::type value_type;
+#endif
```
Comment (eseiler) [inline include/sdsl/iterators.hpp:147]:
```suggestion
    typedef typename std::invoke_result_t<t_F, size_type> value_type;
```

File: include/sdsl/iterators.hpp [hunk 2]
```diff
@@ -9,6 +9,7 @@
 #define INCLUDED_SDSL_ITERATORS
 
 #include <iterator>
+#include <type_traits> // __cpp_lib_is_invocable
```
Comment (eseiler) [inline include/sdsl/iterators.hpp:12]:
```suggestion
#include <type_traits> // std::invoke_result_t
```

**Pull request merged**